test-data/unit/check-enum.test

```diff
@@ -951,3 +951,119 @@ class A:
+else:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+[builtins fixtures/primitives.pyi]
+
+[case testEnumReachabilityWithChainingDisjoint]
+# flags: --warn-unreachable
+from enum import Enum
+
+class Foo(Enum):
+    A = 1
+    B = 2
+
+    # Used to divide up a chained comparison into multiple identity groups
+    def __lt__(self, other: object) -> bool: return True
+
+x: Foo
+y: Foo
+
+# No conflict
+if x is Foo.A < y is Foo.B:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+else:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+# The standard output when we end up inferring two disjoint facts about the same expr
+if x is Foo.A and x is Foo.B:
+    reveal_type(x)  # E: Statement is unreachable
+else:
+    reveal_type(x)  # N: Revealed type is '__main__.Foo'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+
+# TODO: This should behave in the same way as above.
+# However, unlike the above, we currently don't progressively update the type of 'x' as
+# we check each individual comparison. So, when we do 'x is Foo.B', mypy still thinks
+# 'x' is of type 'Foo', which is why we get the below faulty result.
+if x is Foo.A < x is Foo.B:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+else:
+    reveal_type(x)      # N: Revealed type is 'Union[Literal[__main__.Foo.B], Literal[__main__.Foo.A]]'
+reveal_type(x)      # N: Revealed type is 'Union[Literal[__main__.Foo.B], Literal[__main__.Foo.A]]'
+[builtins fixtures/primitives.pyi]
+
+[case testEnumReachabilityWithChainingDirectConflict]
+# flags: --warn-unreachable
+from enum import Enum
+from typing_extensions import Literal, Final
+
+class Foo(Enum):
+    A = 1
+    B = 2
+    C = 3
+
+x: Foo
+if x is Foo.A is Foo.B:
+   reveal_type(x)   # E: Statement is unreachable
+else:
+   reveal_type(x)   # N: Revealed type is '__main__.Foo'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+
+literal_a: Literal[Foo.A]
+literal_b: Literal[Foo.B]
+if x is literal_a is literal_b:
+   reveal_type(x)   # E: Statement is unreachable
+else:
+   reveal_type(x)   # N: Revealed type is '__main__.Foo'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+
+final_a: Final = Foo.A
+final_b: Final = Foo.B
+if x is final_a is final_b:
+   reveal_type(x)   # E: Statement is unreachable
+else:
+   reveal_type(x)   # N: Revealed type is '__main__.Foo'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+
+[builtins fixtures/primitives.pyi]
```

Comment:
Maybe one test with a tricky grouping, like two triples connected with `<`? Ideally you could just add a unit test for the grouping algorithm, but it seems to me this is not easy.

Reply:
Done!

I also refactored out the grouping algorithm into a standalone function and added several unit tests for it to `testinfer.py`. (Not sure is this is the right home for these tests though. LMK if you want me to move them.)